File: 3561-remove-methods-from-project/remove-methods-from-project.java
class Solution{
    public List<Integer> remainingMethods(int n, int k, int[][] invo){
        
        List<List<Integer>> adj=new ArrayList<>();
        for(int i=0; i<n; i++){
            adj.add(new ArrayList<>());
        } 

        for(int[] it: invo){
            adj.get(it[0]).add(it[1]);
        }

        boolean[] visited=new boolean[n];
        // Arrays.fill(visited, false);
        Queue<Integer> q=new LinkedList<>();
        q.offer(k);
        visited[k]=true;

        while(!q.isEmpty()){
            int node=q.poll();

            for(int it: adj.get(node)){
                if(!visited[it]){
                    q.offer(it);
                    visited[it]=true;
                }
            }
        }

        boolean flag=true;
        for(int i=0; i<n && flag; i++){
            if(!visited[i]){
                for(int it: adj.get(i)){
                    if(visited[it]){
                        flag=false;
                        break;
                    }
                }
            }
        }

        List<Integer> res=new ArrayList<>();
        if(!flag){
            for(int i=0; i<n; i++){
                res.add(i);
            }
        }
        else{
            for(int i=0; i<n; i++){
                if(!visited[i]){
                    res.add(i);
                }
            }
        }

        return res;


    }
}

File: 3561-remove-methods-from-project/remove-methods-from-project.py
class Solution:
    def remainingMethods(self, n: int, k: int, invo: List[List[int]]) -> List[int]:

        adj=[[] for _ in range(n)]

        for u, v in invo:
            adj[u].append(v)

        visited=[False]*n
        q=deque([k])
        visited[k]=True

        while q:
            node=q.popleft()

            for it in adj[node]:
                if not visited[it]:
                    q.append(it)
                    visited[it]=True


        flag=True
        for i in range(n):
            if not visited[i]:
                for it in adj[i]:
                    if visited[it]:
                        flag=False
                        break
            if not flag:
                break

        if not flag:
            return list(range(n))

        return [i for i in range(n) if not visited[i]]


        